public/app/features/query/components/QueryEditorRows.tsx

```diff
@@ -67,21 +85,29 @@ export class QueryEditorRows extends PureComponent<Props> {
           {(provided) => {
             return (
               <div ref={provided.innerRef} {...provided.droppableProps}>
-                {queries.map((query, index) => (
-                  <QueryEditorRow
-                    dsSettings={dsSettings}
-                    id={query.refId}
-                    index={index}
-                    key={query.refId}
-                    data={data}
-                    query={query}
-                    onChange={(query) => this.onChangeQuery(query, index)}
-                    onRemoveQuery={this.onRemoveQuery}
-                    onAddQuery={this.props.onAddQuery}
-                    onRunQuery={this.props.onRunQueries}
-                    queries={queries}
-                  />
-                ))}
+                {queries.map((query, index) => {
+                  const dataSourceSettings = getDataSourceSettings(query, dsSettings);
+                  const onChangeDataSourceSettings = dsSettings.meta.mixed
+                    ? (settings: DataSourceInstanceSettings) => this.onDataSourceChange(settings, index)
+                    : undefined;
+
+                  return (
+                    <QueryEditorRow
+                      id={query.refId}
+                      index={index}
+                      key={query.refId}
+                      data={data}
+                      query={query}
+                      dataSourceSettings={dataSourceSettings}
+                      onChangeDataSourceSettings={onChangeDataSourceSettings}
```

Comment:
not sure I like the naming of these, or at lleast onChangeDataSourceSettings feels very misleading, it has nothing to do with changing data source settings. the main QueryEditorRows calls this dsSettings , and the change callback should be onChangeDataSource

Reply:
think that is ok (to check onDataSourceChange)